### PR TITLE
Add `poolSwap` RPC

### LINKED ISSUE #48: DeFiChain Jellyfish RPC Implementations
> This tracks all the RPC methods and categories being implemented in `jellyfish-core`. Will list all the methods here, and strike out those that are not applicable.

### `blockchain.cpp`
- [ ] getblockchaininfo
- [ ] getchaintxstats
- [ ] getblockstats
- [ ] getbestblockhash
- [ ] getblockcount
- [ ] getblock
- [ ] getblockhash
- [ ] getblockheader
- [ ] getchaintips
- [ ] getdifficulty
- [ ] getmempoolancestors
- [ ] getmempooldescendants
- [ ] getmempoolentry
- [ ] getmempoolinfo
- [ ] clearmempool
- [ ] getrawmempool
- [ ] gettxout
- [ ] gettxoutsetinfo
- [ ] pruneblockchain
- [ ] savemempool
- [ ] verifychain
- [ ] preciousblock
- [ ] scantxoutset
- [ ] getblockfilter
- [ ] invalidateblock
- [ ] reconsiderblock
- [ ] waitfornewblock
- [ ] waitforblock
- [ ] waitforblockheight
- [ ] syncwithvalidationinterfacequeue

### `mining.ccp`
* [x] getnetworkhashps
* [x] getmintinginfo
* [ ] prioritisetransaction
* [ ] getblocktemplate
* [ ] submitblock
* [ ] submitheader
* [ ] generatetoaddress
* [ ] estimatesmartfee
* [ ] estimaterawfee

### `misc.cpp`
- [ ] getmemoryinfo
- [ ] logging
- [ ] validateaddress
- [ ] createmultisig
- [ ] deriveaddresses
- [ ] getdescriptorinfo
- [ ] verifymessage
- [ ] signmessagewithprivkey
- [ ] setmocktime
- [ ] echo
- [ ] echojson

### `net.cpp`
- [ ] getconnectioncount
- [ ] ping
- [ ] getpeerinfo
- [ ] addnode
- [ ] disconnectnode
- [ ] getaddednodeinfo
- [ ] getnettotals
- [ ] getnetworkinfo
- [ ] setban
- [ ] listbanned
- [ ] clearbanned
- [ ] setnetworkactive
- [ ] getnodeaddresses

### `rawtransaction.cpp`
- [ ] getrawtransaction
- [ ] createrawtransaction
- [ ] decoderawtransaction
- [ ] decodescript
- [ ] sendrawtransaction
- [ ] combinerawtransaction
- [ ] signrawtransactionwithkey
- [ ] testmempoolaccept
- [ ] decodepsbt
- [ ] combinepsbt
- [ ] finalizepsbt
- [ ] createpsbt
- [ ] converttopsbt
- [ ] utxoupdatepsbt
- [ ] analyzepsbt
- [ ] joinpsbts
- [ ] gettxoutproof
- [ ] verifytxoutproof

### `server.cpp`
- [ ] getrpcinfo
- [ ] help
- [ ] stop
- [ ] uptime

### `spv_rpc.cpp`
- [ ] spv_sendrawtx
- [ ] spv_createanchor
- [ ] spv_createanchortemplate
- [ ] spv_estimateanchorcost
- [ ] spv_rescan
- [ ] spv_syncstatus
- [ ] spv_gettxconfirmations
- [ ] spv_splitutxo
- [ ] spv_listanchors
- [ ] spv_listanchorauths
- [ ] spv_listanchorrewardconfirms
- [ ] spv_listanchorrewards
- [ ] spv_listanchorsunrewarded
- [ ] spv_listanchorspending
- [ ] spv_setlastheight

### `rpcwallet.cpp`
- [ ] fundrawtransaction
- [ ] abandontransaction
- [ ] abortrescan
- [ ] addmultisigaddress
- [ ] backupwallet
- [ ] bumpfee
- [ ] createwallet
- [ ] dumpprivkey
- [ ] dumpwallet
- [ ] encryptwallet
- [ ] getaddressesbylabel
- [ ] getaddressinfo
- [x] getbalance
- [ ] getnewaddress
- [ ] getrawchangeaddress
- [ ] getreceivedbyaddress
- [ ] getreceivedbylabel
- [ ] gettransaction
- [ ] getunconfirmedbalance
- [ ] getbalances
- [ ] getwalletinfo
- [ ] importaddress
- [ ] importmulti
- [ ] importprivkey
- [ ] importprunedfunds
- [ ] importpubkey
- [ ] importwallet
- [ ] keypoolrefill
- [ ] listaddressgroupings
- [ ] listlabels
- [ ] listlockunspent
- [ ] listreceivedbyaddress
- [ ] listreceivedbylabel
- [ ] listsinceblock
- [ ] listtransactions
- [ ] listunspent
- [ ] listwalletdir
- [ ] listwallets
- [ ] loadwallet
- [ ] lockunspent
- [ ] removeprunedfunds
- [ ] rescanblockchain
- [ ] sendmany
- [ ] sendtoaddress
- [ ] sethdseed
- [ ] setlabel
- [ ] settxfee
- [ ] setwalletflag
- [ ] signmessage
- [ ] signrawtransactionwithwallet
- [ ] unloadwallet
- [ ] walletcreatefundedpsbt
- [ ] walletlock
- [ ] walletpassphrase
- [ ] walletpassphrasechange
- [ ] walletprocesspsbt

### `mn_rpc.cpp`
- [ ] setgov
- [ ] getgov
- [ ] isappliedcustomtx

### `rpc_accounts.cpp`
- [ ] listaccounts
- [ ] getaccount
- [ ] gettokenbalances
- [ ] utxostoaccount
- [ ] accounttoaccount
- [ ] accounttoutxos
- [ ] listaccounthistory
- [ ] accounthistorycount
- [ ] listcommunitybalances
- [ ] sendtokenstoaddress

### `rpc_masternodes.cpp`
- [ ] createmasternode
- [ ] resignmasternode
- [ ] listmasternodes
- [ ] getmasternode
- [ ] getmasternodeblocks
- [ ] listcriminalproofs
- [ ] getanchorteams
- [ ] getactivemasternodecount
- [ ] listanchors

### `rpc_poolpair.cpp`
- [ ] listpoolpairs
- [ ] getpoolpair
- [ ] addpoolliquidity
- [ ] removepoolliquidity
- [ ] createpoolpair
- [ ] updatepoolpair
- [ ] poolswap
- [ ] listpoolshares
- [ ] testpoolswap

### `rpc_tokens.cpp`
- [ ] createtoken
- [ ] updatetoken
- [ ] listtokens
- [ ] gettoken
- [ ] getcustomtx
- [ ] minttokens

### PR DESCRIPTION
**Pull request by siradji**
#### What kind of PR is this?:

/kind feature

#### What this PR does / why we need it:

RPC implementation

#### Which issue(s) does this PR fixes?:

Fixes part of #48

### CODE REVIEW

File: website/docs/jellyfish/api/poolpair.md [hunk 1]
```diff
@@ -160,3 +160,48 @@ interface PoolShareOptions {
...
+## testPoolSwap 
+
+Tests a poolswap transaction with given metadata and returns poolswap result. See PoolswapResultType.
+
+```ts title="client.poolpair.testPoolSwap()"
+interface testPoolSwap {
+  poolSwap (metadata: PoolSwapMetadata): Promise<PoolSwapResultType>
+}
+
+export type PoolSwapResultType = `${number}&${string}`
```
Comment (jingyi2811) [inline website/docs/jellyfish/api/poolpair.md:197]: Remove export

File: packages/jellyfish-api-core/__tests__/category/poolpair/testPoolSwap.test.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    await container.waitForWalletCoinbaseMaturity()
+    await container.waitForWalletBalanceGTE(200)
+    await setup()
+  })
+
+  afterAll(async () => {
+    await container.stop()
+  })
+
+  let address: string
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/poolpair/testPoolSwap.test.ts:21]: Move this line to line 29

File: website/docs/jellyfish/api/poolpair.md [hunk 3]
```diff
@@ -160,3 +160,48 @@ interface PoolShareOptions {
...
+interface PoolSwapMetadata {
+  from: string
+  tokenFrom: string
+  amountFrom: number
+  to: string
+  tokenTo: string
+  maxPrice: number
+}
+
+interface PoolSwapInputs {
```
Comment (jingyi2811) [inline website/docs/jellyfish/api/poolpair.md:182]: PoolSwapInputs is a decent name. But for this variable, we declare like below in other typescript file

```
export interface UTXO {
  txid: string
  vout: number
}
```

File: packages/jellyfish-api-core/src/category/poolpair.ts [hunk 4]
```diff
@@ -181,3 +218,17 @@ export interface AddPoolLiquidityUTXO {
...
+export interface PoolSwapMetadata {
+  from: string
+  tokenFrom: string
+  amountFrom: number
+  to: string
+  tokenTo: string
+  maxPrice?: number
+}
+
+export interface PoolSwapInputs {
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/src/category/poolpair.ts:231]: PoolSwapInputs is a decent name. But for this variable, we write this in other typescript file

```
export interface UTXO {
  txid: string
  vout: number
}
```

File: packages/jellyfish-api-core/__tests__/category/poolpair/poolSwap.test.ts [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+    expect(typeof poolSwapHex).toStrictEqual('string')
+    expect(swapAmount).toStrictEqual('33.36113339@1')
+  })
+
+  it('should create a poolswap transaction', async () => {
+    const metadata: PoolSwapMetadata = {
+      from: address, amountFrom: 2, tokenFrom: 'DFI', tokenTo: 'ETH', to: await client.wallet.getNewAddress()
+    }
+    const poolswapHex = await client.poolpair.poolSwap(metadata)
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/poolpair/poolSwap.test.ts:98]: Usually, we do not share 2 RPC in the same file.

**Comment by jingyi2811** [comment]
Thanks for asking me to review. @canonbrother has more experience in poopair. I guess it is related to poolswap / testpoolswap. He probably could help you better

**Comment by siradji** [comment]
@canonbrother , i need you help in reviewing this PR. I'd want you to give me your opinion on the testing. I have a feeling there are more ways to implement more stringent testing. And also, is there a way to test a `poolswap` transaction without using the `testPoolSwap` RPC (refer to the picture below)

image

File: packages/jellyfish-api-core/__tests__/category/poolpair/poolSwap.test.ts [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+      to: await client.wallet.getNewAddress()
+    }
+    const poolswapHex = await client.poolpair.poolSwap(metadata)
+    const swapAmount = await client.poolpair.testPoolSwap(metadata)
+
+    expect(swapAmount).toStrictEqual('0.09910797@0')
+    expect(typeof poolswapHex).toStrictEqual('string')
+  })
+
+  it('should throw an error when invalid token is specified', async () => {
```
Comment (monstrobishi) [inline packages/jellyfish-api-core/__tests__/category/poolpair/poolSwap.test.ts:121]: I would add another test that checks the max price protection. Do this by:

- Fund UTXOs for 2 different addresses
- Sending a tx with max price using first address
- Don't generate a block yet
- Send a second swap tx a big one using second addresss
- Then generate
- Confirm that swap wasn't performed due to max price protection

**Review by monstrobishi — CHANGES_REQUESTED** [review]

**Comment by canonbrother** [comment]
1. test poolswap
  - expect reserveA, reserveB value (refer the calculation below)
  - check before after token balance
2. test poolswap with maxPrice

Calculation for your ref
```
10: 200 = sqrt(10*200) = 44.721359
from: DDAI, fromAmt: 6
? : 206 = sqrt(? * 206) = 44.72135955
? = 44.72135955^2 / 206 = 9.708755427
bal:  [ '90.29126214@0', '1794.00000000@1', '44.72134954@2' ]

from: DFI, fromAmt: 6
16 : ? = sqrt(16 * ?) = 44.72135955
? = 44.72135955^2 / 16 = 125.000225
bal:  [ '84.00000000@0', '1875.00000000@1', '44.72134954@2'
```
3. test poolswap against maxPrice
4. test poolswap against non-existence token
5. test poolswap against poolpair status false

**Comment by siradji** [comment]
> 1. test poolswap (test respectively from diff coin swap)
> 
>   - expect reserveA, reserveB value (refer the calculation below)
> 
>   - check before after token balance
> 
> 2. test poolswap with maxPrice (don't need assert too details as **test poolswap ady done above**, just ensure it should be passed
> 
> 
> 
> Calculation for your reference
> 
> `note: the calculation is based on your poolpair "tokenA: 10, tokenB: 200" set up`
> 
> ```
> 
> beforeBal:  [ '90.00000000@0', '1800.00000000@1', '44.72134954@2' ]
> 
> 10: 200 = sqrt(10*200) = 44.721359
> 
> 
> 
> from: DDAI, fromAmt: 6
> 
> ? : 206 = sqrt(? * 206) = 44.72135955
> 
> ? = 44.72135955^2 / 206 = 9.708755427
> 
> afterBal:  [ '90.29126214@0', '1794.00000000@1', '44.72134954@2' ]
> 
> 
> 
> from: DFI, fromAmt: 6
> 
> 16 : ? = sqrt(16 * ?) = 44.72135955
> 
> ? = 44.72135955^2 / 16 = 125.000225
> 
> afterBal:  [ '84.00000000@0', '1875.00000000@1', '44.72134954@2'
> 
> ```
> 
> 3. test poolswap against maxPrice
> 
> 4. test poolswap against non-existence token
> 
> 5. test poolswap against poolpair status false
> 
> 
> 
>     

@canonbrother , the calculation is a little bit complex. I would appreciate it if you can clarify further.

**Pull request closed without merging**